consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/validation/MessageValidatorTest.java

```diff
@@ -0,0 +1,274 @@
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.IbftSignedMessageData;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.IbftUnsignedCommitMessageData;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.IbftUnsignedPrePrepareMessageData;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.IbftUnsignedPrepareMessageData;
+import tech.pegasys.pantheon.crypto.SECP256K1;
+import tech.pegasys.pantheon.crypto.SECP256K1.KeyPair;
+import tech.pegasys.pantheon.ethereum.ProtocolContext;
+import tech.pegasys.pantheon.ethereum.chain.MutableBlockchain;
+import tech.pegasys.pantheon.ethereum.core.Address;
+import tech.pegasys.pantheon.ethereum.core.Block;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.core.Hash;
+import tech.pegasys.pantheon.ethereum.core.Util;
+import tech.pegasys.pantheon.ethereum.db.WorldStateArchive;
+import tech.pegasys.pantheon.ethereum.mainnet.BlockHeaderValidator;
+
+import java.util.List;
+
+import com.google.common.collect.Lists;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class MessageValidatorTest {
+
+  private final KeyPair proposerKey = KeyPair.generate();
+  private final KeyPair validatorKey = KeyPair.generate();
+  private final KeyPair nonValidatorKey = KeyPair.generate();
+  private final IbftMessageFactory proposerMessageFactory = new IbftMessageFactory(proposerKey);
+  private final IbftMessageFactory validatorMessageFactory = new IbftMessageFactory(validatorKey);
+  private final IbftMessageFactory nonValidatorMessageFactory =
+      new IbftMessageFactory(nonValidatorKey);
+
+  private List<Address> validators = Lists.newArrayList();
+
+  @Mock private BlockHeaderValidator<IbftContext> headerValidator;
+  private BlockHeader parentHeader = mock(BlockHeader.class);
+  private final ConsensusRoundIdentifier roundIdentifier = new ConsensusRoundIdentifier(2, 0);
+  private MessageValidator validator;
+
+  private final Block block = mock(Block.class);
+
+  @Before
+  public void setup() {
+    validators.add(Util.publicKeyToAddress(proposerKey.getPublicKey()));
+    validators.add(Util.publicKeyToAddress(validatorKey.getPublicKey()));
+
+    final ProtocolContext<IbftContext> protocolContext =
+        new ProtocolContext<>(
+            mock(MutableBlockchain.class), mock(WorldStateArchive.class), mock(IbftContext.class));
+
+    validator =
+        new MessageValidator(
+            validators,
+            Util.publicKeyToAddress(proposerKey.getPublicKey()),
+            roundIdentifier,
+            headerValidator,
+            protocolContext,
+            parentHeader);
+
+    when(block.getHash()).thenReturn(Hash.fromHexStringLenient("1"));
+  }
+
+  @Test
+  public void receivingAPrepareMessageBeforePrePrepareFails() {
```

Comment:
The names of these tests suggest that if we receive a Prepare or Commit message before we receive a Pre-prepare message then the Prepare/Commit message should be rejected. Perhaps there is a better name for these tests.

Reply:
For the context, this is the expected behaviour - i.e. this class will state that a Prepare message is invalid if insufficient context exists to validate it.